src/dune_rules/warning.ml

```diff
@@ -0,0 +1,78 @@
+open Import
+
+module Name = struct
+  include String
+
+  include Stringlike.Make (struct
+      type t = string
+
+      let module_ = "Warning"
+      let description = "Warning name"
+      let description_of_valid_string = None
+      let hint_valid = None
+      let of_string_opt x = if Dune_sexp.Atom.is_valid x then Some x else None
+      let to_string x = x
+    end)
+end
+
+type t =
+  { name : Name.t
+  ; since : Syntax.Version.t
+  ; default : Syntax.Version.t -> Config.Toggle.t
+  }
+
+let frozen = ref false
+let all : (Name.t, t) Table.t = Table.create (module Name) 12
+
+let create ~default ~name ~since =
```

Comment:
I think we can remove the need to dynamically build up a table of warnings which would simplify the code and remove the possibility of hitting the `"Warnings may not be created after we've read them"` error. I suggest we make this `create` function private which would require that all warnings be defined centrally in this file. Then we could statically create a list in this file containing all the warnings.

Reply:
Eventually, it will be possible for extensions to define their own warnings. E.g. ctypes, melange, coq should all be able to do so. For now, I'm fine with keeping it this way.

Although I don't see how we can remove the table even if everything is in one place. E.g. we'll have something like:
```
let w1 = ..
let w2 = ..
let all = [w1 ; w2 ; .. ]
```

And we'll need a check that no warnings were created that aren't inside of `all`.